engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+// cancellation allows a clean exit when needed.
+// NOTE: this is a destructive operation. Check permissions before applying!
+// TODO: move the AllowDeletion policy inside this function for safety?
+func (obj *HetznerVMRes) deleteServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("deleteServer()")
+	}
+	if obj.server == nil {
+		return fmt.Errorf("server is already unavailable")
+	}
+	if _, err := obj.client.Server.Delete(ctx, obj.server); err != nil {
+		return errwrap.Wrapf(err, "client.server.delete failed")
+	}
+	if err := obj.waitFor(ctx, serverAbsent); err != nil {
+		return errwrap.Wrapf(err, "waitFor(serverAbsent) exited early")
+	}
+	return nil
+}
+
+// rebuildServer deletes the current server instance and creates a new one, in
+// accordance with the provided resource specifications.
+// NOTE: this is a destructive operation, don't call without AllowDeletion.
+func (obj *HetznerVMRes) rebuildServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("rebuildServer()")
+	}
+	if obj.server == nil {
+		return fmt.Errorf("server is unavailable")
+	}
+	if err := obj.deleteServer(ctx); err != nil {
+		return errwrap.Wrapf(err, "deleteServer failed")
+	}
+	if err := obj.createServer(ctx); err != nil {
+		return errwrap.Wrapf(err, "createServer failed")
+	}
+	return nil
+}
+
+// getServerConfig converts resource fields to hcloud.ServerCreateOpts struct,
+// such that a server instance created with those Opts has the desired specs.
+func (obj *HetznerVMRes) getServerConfig(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getServerConfig()")
+	}
+	// default, volumes not supported (yet)
+	automount := false
+	// poweron at creation to reach `running` faster
+	startAfterCreate := (obj.State == StateRunning)
+	// build serverconfig elements
+	serverType, _, err := obj.client.ServerType.GetByName(ctx, obj.ServerType)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect ServerType struct")
+	}
+	image, _, err := obj.client.Image.GetByName(ctx, obj.Image)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect Image struct")
+	}
+	datacenter, _, err := obj.client.Datacenter.GetByName(ctx, obj.Datacenter)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect Datacenter struct")
+	}
+	// TODO: add more flexible key selection
+	keylist, err := obj.client.SSHKey.All(ctx)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect SSHKey array")
+	}
+	// build serverconfig from given specs & defaults
+	obj.serverconfig = hcloud.ServerCreateOpts{
+		Name:             obj.Name(),        // string
+		ServerType:       serverType,        // *ServerType
+		Image:            image,             // *Image
+		SSHKeys:          keylist,           // []*SSHKey
+		Location:         nil,               // *Location
+		Datacenter:       datacenter,        // *Datacenter
+		UserData:         "",                // string
```

Comment:
This might be the correct way to kick-off mgmt in the new machine... Look at the other userdata in https://github.com/purpleidea/mgmt/blob/master/engine/resources/aws_ec2.go#L166
I think we might want to have this in the main struct or is it something else?

Reply:
If we just need to accept a string param to make this work it's no problem to add that feature. I'm not familiar with how to use it, but from the aws docs it does look more useful than I thought. Thanks for pointing this out.